utils/joiner/tuplejoiner.cpp

```diff
@@ -199,6 +235,166 @@ bool TupleJoiner::operator<(const TupleJoiner& tj) const
     return size() < tj.size();
 }
 
+void TupleJoiner::getBucketCount()
+{
+    // get the # of cores, round up to nearest power of 2
+    // make the bucket mask
+    numCores = sysconf(_SC_NPROCESSORS_ONLN);
+    if (numCores <= 0)
+        numCores = 8;
+    bucketCount = (numCores == 1 ? 1 : (1 << (32 - __builtin_clz(numCores - 1))));
+    bucketMask = bucketCount - 1;
+}
+
+template<typename buckets_t, typename hash_table_t>
+void TupleJoiner::bucketsToTables(buckets_t *buckets, hash_table_t *tables)
+{
+    uint i;
+
+    bool done = false, wasProductive;
+    while (!done)
+    {
+        done = true;
+        wasProductive = false;
+        for (i = 0; i < bucketCount; i++)
+        {
+            if (buckets[i].empty())
+                continue;
+            bool gotIt = m_bucketLocks[i].try_lock();
+            if (!gotIt)
+            {
+                done = false;
+                continue;
+            }
+            for (auto &element : buckets[i])
+                tables[i]->insert(element);
+            m_bucketLocks[i].unlock();
+            wasProductive = true;
+            buckets[i].clear();
+        }
+        if (!done && !wasProductive)
+            ::usleep(1000 * numCores);
```

Comment:
This potentially leads to an alignment of separate threads with a skewed data set. A non-linear delay with a smaller constant could be of use or use ASM instruction sequence number as a random value. But the latter sounds like an overkill.

Reply:
I think I see what you're getting at.  There's already a good deal of randomness in the system though.  But true, threads could consistently block on a few # of a buckets or even a single bucket, and bucket(s) may not be fully utilized b/c of the sleeps.  I'm going to test performance when inserting millions of duplicates to see if there's a regression there.  Will mod if there's a substantial performance loss.

Reply:
I might use inapproriate term. There could be a potential syncronization of threads with a skewed data set so they will gather together waiting for the lock.